### DISCUSSION TITLE
How to safely disConnect?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8152] [Q&A]
Hello, I'm using JNI to make RDP functionality available on the web. ( Not Android )

Is there a way to safely disconnect from a simple client using JNI like an Android client?

A simple trigger, i was making to stop the main loop's while statement when calling the currently created disconnect (JNI) function.

I also tried to run related functions recklessly.

`freerdp_client_stop` , `freerdp_disconnect` ,` freerdp_abort_connect_context` , ......

After main loop stops, freerdp_disconnect and freerdp_client_stop are called like other clients , but I don't know if the memory is properly cleaned up and completely free.

An error is also called when stopping.

`[ERROR][com.freerdp.core] - freerdp_abort_connect_context:freerdp_set_last_error_ex ERRCONNECT_CONNECT_CANCELLED [0x0002000B]`

`SetEvent()` `CreateEvent()` 
I tried to deliver the stop event using this, but I don't know if it applies to the sample client.

Please advise if there is a way to safely end the loop of the sample client and completely `free` the context related to the instance.

How do I using jni to adjust `freerdp_shall_disconnect()` to get true in main loop?

Thank you for reading the question.

**Comment by akallabeth** [comment]
1. look at the sample clients on how the main loop is done for resource allocation/cleanup (check the pre/post connect/disconnect callbacks, ...)
2. yes, you do that right. calling `freerdp_abort_connect_context` from any thread will terminate the main thread. after that the context/instance needs to be cleaned up like with the sample clients.
3. Create/extend the `JNI` any way you like to get that properly working (but the current android client `JNI` should suffice)
